tests/objects/test_world_object.py

```diff
@@ -284,36 +284,3 @@ def test_axis_setters():
 
     obj.world.up = (1, 1, 1)
     assert np.allclose(obj.world.up, (1 / np.sqrt(3), 1 / np.sqrt(3), 1 / np.sqrt(3)))
-
-
-def test_gravity():
```

Comment:
This whole test was removed. Would it not still be valuable, if the names were fixed?

Reply:
The test was checking that `reference_up` is synced correctly between transforms, i.e., if I set `ob.local.reference_up = (1, 1, 1)` then it checked that `obj.world.reference_up` also becomes `la.vector_apply_matrix(ob.local.reference_up, ob.world.parent.matrix)`. And vice versa.

The implementation had a bug/side-effect where `reference_up` was now synced scene-wide, because `obj.world.reference_up` used the `reference_up` of its parent. I initially fixed it by giving `RecursiveTransform` its own `reference_up` which is defined in the same frame as `RecursiveTransform.parent` and synced to `RecursiveTransform.own.reference_up`.

Then I realized that the direction of the sync is also opinionated. If we sync `RecursiveTransform -> AffineTransform` (`obj.world.up_reference` is true) then `obj.world.up_reference` is preserved when the parent updates. While we want this most of the time, we also want to allow scenes where the camera follows a person/object walking on a sphere, the hull of a spaceship, or another curved surface where `up_reference` should be in sync with the surface normal of the object being walked on. It also hardcodes the assumption that an `AffineTransform` is wrapped/used by at most one `RecursiveTransform`. This is okay for now (WorldObjects are locked to a single parent), but if we ever want to allow things like re-using objects in different/multiple scenes then this is another place where patching is needed. If we sync `AffineTransform -> RecursiveTransform` then `ob.world.up_reference` will change whenever `ob.parent.world.transform` changes and that is not what we want most of the time.

Long story short, I decided to remove the sync and thus removed the test checking if the sync works. 

---

Thinking about this again after writing the reply I think we could actually go with the `RecursiveTransform -> AffineTransform` sync and then ask people with the "walking on spheres" use-case to register a callback on the world transform of the object being followed that sets the camera's `up_reference` whenever the object updates, i.e. something like

```python
def ground_camera(transform):
    # object stands on curved surface, its up is the cameras up reference
    camera.world.up_reference = tracked_obj.world.up

    # or with local references only
    new_ref = la.vector_apply_matrix((0, 1, 0), transform)
    camera.world.up_reference = new_value

tracked_obj.world.on_update(ground_camera)
```

I can have a look at bringing the test back and re-introducing the sync. Alternatively, we could also drop `ob.local.up_reference` for now and only have a `ob.world.up_reference` property. We currently don't use `ob.local.up_reference` so it is possible to skip. Thoughts? @almarklein @Korijn

Reply:
Resolved. objects now have independent `reference_up` values that are synced between `obj.world` and `obj.local` and are both given in their respective target frame, i.e., `obj.world.reference_up` is given in world coordinates and `obj.local.reference_up` is given in parent coordinates.